fish-rust/src/builtins/shared.rs

```diff
@@ -1,64 +1,66 @@
-use crate::builtins::{printf, wait};
-use crate::common::str2wcstring;
-use crate::ffi::separation_type_t;
-use crate::ffi::{self, parser_t, wcstring_list_ffi_t, Repin, RustBuiltin};
+use super::prelude::*;
+use crate::builtins::*;
+use crate::common::{escape, get_by_sorted_name, read_blocked, str2wcstring, Named};
+use crate::ffi;
+use crate::io::{IoChain, IoFd, OutputStream, SeparationType};
+use crate::parse_constants::UNKNOWN_BUILTIN_ERR_MSG;
+use crate::parse_util::parse_util_argument_is_help;
+use crate::parser::{Block, BlockType, LoopStatus};
+use crate::proc::{no_exec, ProcStatus};
 use crate::wchar::{wstr, WString, L};
-use crate::wchar_ffi::{c_str, empty_wstring, ToCppWString, WCharFromFFI};
+use crate::wchar_ffi::{c_str, empty_wstring, ToCppWString, W0String, WCharFromFFI, WCharToFFI};
 use crate::wgetopt::{wgetopter_t, wopt, woption, woption_argument_t};
+use crate::wutil::perror;
 use cxx::{type_id, ExternType};
-use libc::c_int;
 use libc::isatty;
-use libc::STDOUT_FILENO;
+use libc::{c_int, STDERR_FILENO, STDIN_FILENO, STDOUT_FILENO};
 
-use std::borrow::Cow;
+use errno::errno;
 use std::fs::File;
 use std::io::{BufRead, BufReader, Read};
 use std::os::fd::{FromRawFd, RawFd};
 use std::pin::Pin;
+use std::sync::Arc;
+use widestring_suffix::widestrs;
 
-pub type BuiltinCmd = fn(&mut parser_t, &mut io_streams_t, &mut [&wstr]) -> Option<c_int>;
-
-#[cxx::bridge]
-mod builtins_ffi {
-    extern "C++" {
-        include!("wutil.h");
-        include!("parser.h");
-        include!("builtin.h");
-
-        type wcharz_t = crate::ffi::wcharz_t;
-        type wcstring_list_ffi_t = crate::ffi::wcstring_list_ffi_t;
-        type parser_t = crate::ffi::parser_t;
-        type io_streams_t = crate::ffi::io_streams_t;
-        type RustBuiltin = crate::ffi::RustBuiltin;
-    }
-    extern "Rust" {
-        fn rust_run_builtin(
-            parser: Pin<&mut parser_t>,
-            streams: Pin<&mut io_streams_t>,
-            cpp_args: &wcstring_list_ffi_t,
-            builtin: RustBuiltin,
-            status_code: &mut i32,
-        ) -> bool;
-    }
-}
+pub type BuiltinCmd = fn(&Parser, &mut IoStreams<'_>, &mut [WString]) -> Option<c_int>;
```

Comment:
https://github.com/fish-shell/fish-shell/commit/d6b41b33f4c888bbce3cba1e14e3dbf9dadc0afd#r124997196
> I wrote: I disagree with the changing of `&mut [&wstr]` to `&mut [WString]`. It is _very_ noisy and I don't see a good reasoning for it. In my opinion `&wstr` is actually a more useful type since we essentially never want to modify it directly. Utilising any spare capacity or mutating in-place seems like premature optimisations.

> @krobelus responded: 
> I changed [&wstr] to [WString] (I think I mentioned it on your PR before) because unlike [&wstr], it occurs naturally in our code.
> It seemed more elegant to not require conversion for builtin execution.
> My suspicion was that the existing Rust uses [&wstr] because it didn't account for the the whole code base so I thought I was correcting an oversight here.. which is unlikely I guess.
> 
> &wstr is actually a more useful type since we essentially never want to modify it directly.
> This property we could probably encode independent of whether we use a reference.
> 
> I might very well be wrong. You mention some good points in https://matrix.to/#/!YLTeaulxSDauOOxBoR:matrix.org/$B2oLWotpXrSVble7xAPpoV_mi-4FNA6n2BwSH46l8lE?via=gitter.im&via=matrix.org&via=tchncs.de
> My opinion is not very strong here. I'm happy to defer to whoever takes ownership of this code.
> 
> I wonder if we could get the best of both worlds by taking something like [AsRef<wstr>].. however I don't want to invest too much time in this relatively minor issue (it's pretty much isolated to the wgetopt module), I'd rather get the rest of the system working. There's a whole lot to clean up

Reply:
> I wonder if we could get the best of both worlds by taking something like [AsRef].. however I don't want to invest too much time in this relatively minor issue (it's pretty much isolated to the wgetopt module), I'd rather get the rest of the system working. There's a whole lot to clean up

My two cents: make `wgetoptr_t` take `AsRef<wstr>`, extract what you changed about it to a separate PR.

If we want to change the `BuiltinCmd`-type to use `WString`, that should be a separate PR that should land before this (even if that requires cloning to support it until this lands, since it is so breaking). I still disagree with changing it from `&wstr` and I would rather make the private encoding optional and use `BStr` aka `Vec<u8>`, but that's a very different discussion I'd rather do after the port is finished